packages/loopring_v3/contracts/impl/libexchange/ExchangeWithdrawals.sol

```diff
@@ -283,44 +284,56 @@ library ExchangeWithdrawals
         uint amount = (data & 0xFFFFFFF).decodeFloat();
 
         if (amount > 0) {
-            // Set everything to 0 for this withdrawal so it cannot be used anymore
-            data = data & uint(~((1 << (7 * 8)) - 1));
-
-            // Update the data in storage
-            if (withdrawBlock.withdrawals.length >= 32) {
-                assembly {
-                    mstore(add(slice, offset), data)
-                }
-                uint dataSlot1 = slice[0];
-                uint dataSlot2 = slice[1];
-                assembly {
-                    sstore(add(sc, slot1), dataSlot1)
-                    sstore(add(sc, slot2), dataSlot2)
-                }
+            address recipient = S.accounts[accountID].owner;
+            // Transfer the tokens
+            if (useGasLimit) {
+                success = sendTokens(
+                    S,
+                    recipient,
+                    tokenID,
+                    amount
+                );
             } else {
-                bytes memory mWithdrawals = withdrawBlock.withdrawals;
-                assembly {
-                    mstore(add(mWithdrawals, offset), data)
-                }
-                withdrawBlock.withdrawals = mWithdrawals;
+                success = transferTokens(
+                    S,
+                    recipient,
+                    tokenID,
+                    amount
+                );
```

Comment:
should we create a wrapping method for code between line 289 and 302, then inside the method emit a TokenDistributionFailure event if the `success` is false?

Reply:
Added a `WithdrawalFailed`, which can only be emitted while distributing. When a user manually withdraws the transaction will fail with `TRANSFER_FAILURE` when something goes wrong with the token transfer.